db/commitlog/commitlog.cc

```diff
@@ -1881,6 +1909,9 @@ void db::commitlog::segment_manager::on_timer() {
         if (cfg.mode != sync_mode::BATCH) {
             sync();
         }
+
+        _sync_semaphore.signal(_sync_semaphore.waiters());
+
```

Comment:
Adding to the conversation from Dec 2020: I think it's better for the timer to wake a sleeping coroutine (or do_until()) loop. This gets rid of the gate, and serializes timer invocations. The main service can wait for that coroutine to complete and so avoid use-after-frees. For a future cleanup of course.